fastjet/internal/delaunay/triangulation.go

```diff
@@ -0,0 +1,1066 @@
+	}
+	d.triangles = make([]*Triangle, 0)
+	for _, p := range points {
+		d.InsertPoint(p)
+	}
+	return d
+}
+
+// WalkDelaunay creates a delaunay triangulation with the given points
+// it uses the remembering stochastic walk method to find the triangle in which p is inserted
+// it has a worst time complexity of O(n^5/3) but it doesn't need bounds
+func WalkDelaunay(points []*Point) *Delaunay {
+	// it needs at least one triangle to start with
+	if len(points) < 3 {
+		panic(fmt.Errorf("delaunay: not enough points"))
+	}
+	rand.Seed(time.Now().UnixNano())
+	d := &Delaunay{
+		hierarchy: false,
+	}
+	d.triangles = make([]*Triangle, 1)
+	// create first triangle
+	d.triangles[0] = NewTriangle(points[0], points[1], points[2])
+	d.triangles[0].inD = true
+	points[0].adjacentTriangles = points[0].adjacentTriangles.appendT(d.triangles[0])
+	points[1].adjacentTriangles = points[1].adjacentTriangles.appendT(d.triangles[0])
+	points[2].adjacentTriangles = points[2].adjacentTriangles.appendT(d.triangles[0])
+	for i := 3; i < len(points); i++ {
+		d.InsertPoint(points[i])
+	}
+	return d
+}
+
+// Triangle returns all delaunay Triangles
+func (d *Delaunay) Triangles() []*Triangle {
+	if d.hierarchy {
+		// remove triangles that contain the root points
+		rt := make(triangles, len(d.root.A.adjacentTriangles)+len(d.root.B.adjacentTriangles)+len(d.root.C.adjacentTriangles))
+		n := copy(rt, d.root.A.adjacentTriangles)
+		n += copy(rt[n:], d.root.B.adjacentTriangles)
+		copy(rt[n:], d.root.C.adjacentTriangles)
+		return d.triangles.finalize(rt...)
+	}
+	return d.triangles.finalize()
+}
+
+func (d *Delaunay) InsertPoint(p *Point) {
+	p.adjacentTriangles = make(triangles, 0)
+	var t *Triangle
+	var onE bool
+	if d.hierarchy {
+		t, onE = findTriangle(d.root, p)
+	} else {
+		var l, r *Point
+		var start *Triangle
+		// find a triangle to start the walk
+		// use the last changed triangle to start from
+		for i := len(d.triangles) - 1; ; i-- {
+			if d.triangles[i].inD {
+				start = d.triangles[i]
+				break
+			}
+		}
+		t, onE, l, r = walkTriangle(start, p)
+		if t == nil {
+			// point is on the outside
+			d.addPoint(p, l, r)
+			return
+		}
+	}
+	if t == nil {
+		panic(fmt.Errorf("delaunay: no triangle which contains P%v. Min and Max values must be wrong.", p))
+	}
+	if onE {
+		d.insertPonE(t, p)
+	} else {
+		d.insertP(t, p)
+	}
+}
+
+// addPoint adds point on the outside. It checks the neighbors of l and r and creates
+// triangles between p,l,r and the neighbors if they can reach p without crossing any lines
+func (d *Delaunay) addPoint(p, l, r *Point) {
+	// need to find points next to l and r on the outside and check if there orientation<0 add triangles with all of those points and l and r
+	// r is counterclockwise of l
+	// so l's clockwise neighbor needs to be checked
+	var outer1, outer2 []*Point
+	previous1, previous2 := r, l
+	outer1 = append(outer1, l)
+	outer2 = append(outer2, r)
+	found1, found2 := true, true
+	for i := 0; found1; i++ {
```

Comment:
It looks to me like this logic can be factored out into a helper for the two points.